common/djangoapps/course_modes/views.py

```diff
@@ -101,10 +101,18 @@ def get(self, request, course_id, error=None):
                     redirect_url = ecommerce_service.checkout_page_url(professional_mode.bulk_sku)
             return redirect(redirect_url)
 
+        # If the learner has arrived at this screen via the traditional enrollment workflow, then they should
+        # already be enrolled in an audit mode for the course, assuming one has been configured.  However,
+        # alternative enrollment workflows have been introduced into the system, such as third-party discovery.
+        # These workflows result in learners arriving directly at this screen, and they will not necessarily
+        # be pre-enrolled in the audit mode.  In that case we'll need to enroll them here before going further
+        # and then reload the state of the "enrollment_mode" and "is_active" parameters.
+        if not is_active and (not enrollment_mode or enrollment_mode == CourseMode.AUDIT):
```

Comment:
Yeah, I went back and forth on this a bit.  If a learner has an inactive non-audit/premium (eg, verified) enrollment on their record, should they be auto-enrolled in the audit mode assuming one exists?  The current behavior in master just sends you right along to the dashboard, and I didn't want to mess w/ that flow, because I don't have enough context to know if there's a valid scenario.  Will try to catch up with @griffresch and @marcotuts about it.